[MRG+1] Scikit - Optimize based GridSearchCV plug-in replacement

iaroslav-ai opened the pull request:
A class as discussed in #78 under the tentative name of `SkoptSearchCV` (discussion on what might be a better name is welcome :) 

Minimalist usage example as of right now:

```python
from skopt.wrappers import SkoptSearchCV
from skopt.space import Real, Categorical, Integer

opt = SkoptSearchCV(
    SVC(),
    [{
        'C': Real(1e-6, 1e+6, prior='log-uniform'),
        'gamma': Real(1e-6, 1e+1, prior='log-uniform'),
        'degree': Integer(1, 8),
        'kernel': Categorical(['linear', 'poly', 'rbf']),
    }],
    n_jobs=1, n_iter=32,
)

opt.fit(X_train, y_train)
print(opt.score(X_test, y_test))
```

If something is missing in todos let me know. 

- [x] add `skopt.wrappers` to `setup.py`
- [x] implement basic wrapper similar to `RandomizedSearchCV` using `BaseSearchCV`
- [x] support multiple search spaces as list of dicts, similar to `GridSearchCV`
- [x] support parallel search 
- [x] add draft of example usage in `sklearn-wrapper`
- [x] add all necessary data to `cv_results_`
- [x] add tests
- [x] fix the docstrings and add comments 
- [x] review and improve example usage

iaroslav-ai commented:
Also let me know if you do not like the location of the class (cuttently `skopt.wrappers`)

iaroslav-ai commented:
As of right now PR contains working Implementation of `*SearchCV` according to #78 and scikit-learn/scikit-learn#5491 .
Could someone review? I am not sure who would be the best to ask to do this, as all of you guys seemed to be pretty active about this in referenced issue and PR :)

betatim commented:
Thanks for this! Code, example, and tests all in one!

> Also let me know if you do not like the location of the class (cuttently `skopt.wrappers`)

I'm not mad keen on the name. I'd make it available at the top level like `*_minimize`. But then `skopt.SkoptSearchCV` is a bit doppelt gemoppelt. `SMBOSearchCV` is cryptic, ... will think a bit and see what others have used. I think once we know a good name we can use it for the class and replace `wrappers` with it as well.

betatim commented:
[Looks like we need to be smarter about our tests](https://travis-ci.org/scikit-optimize/scikit-optimize/jobs/244194986) so that they can run in less time.

betatim commented on skopt/wrappers/search_cv.py:
```diff
@@ -0,0 +1,490 @@
+import sklearn.model_selection._search as skms
+
+import numpy as np
+from collections import *
```
no stars please :)

glouppe commented on skopt/tests/test_searchcv.py:
```diff
@@ -0,0 +1,80 @@
+
+# Extract available surrogates, so that new ones are used automatically
+available_surrogates = [
+    getattr(sol, name) for name in sol.__all__
```
I would rather  be explicit and write down the full list of surrogates, instead of relying on some black magic that could easily break down as we add/update things.

glouppe commented on skopt/tests/test_searchcv.py:
```diff
@@ -0,0 +1,80 @@
+    """
+    Tests whether the cross validation search wrapper around sklearn
+     models runs properly with available surrogates and with single
+     or multiple workers.
```
indentation issue

glouppe commented on skopt/wrappers/__init__.py:
```diff
@@ -0,0 +1 @@
+from .search_cv import SkoptSearchCV
```
missing line break

glouppe commented on skopt/wrappers/search_cv.py:
```diff
@@ -0,0 +1,490 @@
+        dimensions = [params_space[k] for k in sorted(params_space.keys())]
+
+        if self.surrogate == "auto":
+            surrogate = GaussianProcessRegressor()
```
Is that a good kernel? 
I think it would be nice to solve #338 to reuse a sensible default here.

betatim replied:
We should solve https://github.com/scikit-optimize/scikit-optimize/issues/400#issuecomment-309020353 or we need to add the same kernel setup and space conversion as is in `gp_minimize`

iaroslav-ai replied:
Yup I use now strings for surrogate for Optimizer which cooks me estimators like a chef :stuck_out_tongue_winking_eye:

glouppe commented on skopt/wrappers/search_cv.py:
```diff
@@ -0,0 +1,490 @@
+
+from sklearn.base import clone
+from sklearn.externals.joblib import Parallel, delayed, cpu_count
+import sklearn.model_selection._search as skms
```
can we find a better name than `skms`?

betatim commented on skopt/tests/test_searchcv.py:
```diff
@@ -0,0 +1,80 @@
+# Extract available surrogates, so that new ones are used automatically
+available_surrogates = [
+    getattr(sol, name) for name in sol.__all__
+    if "GradientBoostingQuantileRegressor" not in name
```
Why can't we use this model?

betatim commented on skopt/wrappers/search_cv.py:
```diff
@@ -0,0 +1,490 @@
+        number of evaluations set to self.n_iter. Alternatively, if
+        a list of (dict, int > 0) is given, the search is done for
+        every search space for number of iterations given as a second
+        element of tuple.
```
This means if I give a list of three dictionaries there will be a total of `3*n_iter` iterations? What is scikit-learn's behaviour wrt this?

I thought sklearn treats the situation of having a list of `dict`s as having an extra implied categorical dimension.

We should try and be as much plug'n'play as possible.

iaroslav-ai replied:
The documentation for `GridSearchCV` says that if the list of dicts is provided, then grid spanned by every dict is explored sequentially. Hence I do the same in here.

betatim commented on skopt/wrappers/search_cv.py:
```diff
@@ -0,0 +1,490 @@
+
+        # this dict is used in order to keep track of skopt Optimizer
+        # instances for different search spaces (str(space) is used as key)
+        self.optimizer = {}
```
-> `_optimizer`? It is 'private' to the internal process no?

betatim commented on skopt/wrappers/search_cv.py:
```diff
@@ -0,0 +1,490 @@
+
+            # if tuple: (dict: search space, int: n_iter)
+            if isinstance(elem, tuple):
+                psp, n_iter = elem
```
`psp` -> `search_space` (I think) in general we should use proper words for the variable names instead of abbreviations

betatim commented on skopt/wrappers/search_cv.py:
```diff
@@ -0,0 +1,490 @@
+                psp, n_iter = elem, self.n_iter
+            else:
+                raise ValueError("Unsupported type of parameter space. "
+                                 "Expected tuple or dict, got " + str(elem))
```
switch to `"got %s." % (elem)`

betatim commented on skopt/wrappers/search_cv.py:
```diff
@@ -0,0 +1,490 @@
+                                 "Expected tuple or dict, got " + str(elem))
+
+            # do the optimization for particular search space
+            while n_iter:
```
`bool(-3) == True` so this won't stop if `n_iter` isn't dividable by `n_jobs`. I think we should check if there are enough iterations left to sample `n_jobs` more points, if not reduce it to the number left and then stop the loop. (and add a unit test to check this works)

betatim commented on skopt/wrappers/search_cv.py:
```diff
@@ -0,0 +1,490 @@
+        }
+        return params_dict
+
+    def step(self, X, y, param_space, groups=None, n_jobs=1):
```
I'd have to read the code for `RandomSearchCV` but doesn't it have a step function we can re-use/hook into so we don't have to duplicate all the code for recording the results?

iaroslav-ai commented:
Addressed most of comments. I use now `_fit` method from `BaseSearchCV` inside step function, which simplifies stuff a bit, but might be a bit hacky (because _fit resets results_cv_). Also I renamed the class to BayesSearchCV, as per discussion in scikit-learn/scikit-learn#5491 . If someone has an idea for better name let me know.

iaroslav-ai commented:
Just double checked - this should address all the comments. Reduced the tests a bit - only one estimator class is tested, but generalization to other estimators is tested with the consistency of estimators interface in other tests. Also more general `optimizer_kwargs` is used instead of `base_estimator`.

iaroslav-ai commented:
Please take another look and tell me what you think.

betatim commented on skopt/searchcv.py:
```diff
@@ -0,0 +1,448 @@
+class BayesSearchCV(sk_model_sel.BaseSearchCV):
+    """Bayesian optimization over hyper parameters.
+
+    SkoptSearchCV implements a "fit" and a "score" method.
```
-> BayesSearchCV

iaroslav-ai replied:
Ooops, will fix :D

betatim commented on skopt/searchcv.py:
```diff
@@ -0,0 +1,448 @@
+
+        # this dict is used in order to keep track of skopt Optimizer
+        # instances for different search spaces (str(space) is used as key)
+        self.optimizer_ = {}
```
I like this as a way to support sub-search spaces (link #108 for reference). Maybe we can use this idea to build #108.

betatim commented on skopt/searchcv.py:
```diff
@@ -0,0 +1,448 @@
+
+        return optimizer
+
+    def _skopt_to_dict(self, params_space, params):
```
should we make this guy available in `skopt.utils`? Was discussed in #140 and might be a useful tool in general for people building things themselves?

betatim replied:
(together with the inverse function?)

iaroslav-ai replied:
I agree. Will put it in there somewhere.

betatim commented on skopt/searchcv.py:
```diff
@@ -0,0 +1,448 @@
+
+    def _skopt_to_dict(self, params_space, params):
+        """Converts list of parameter values into the dictionary with
+        keys as parameter names and corresponding values of parameter.
```
Would change to:
"""
Convert a list of parameter values into a dictionary.

Parameter names are used as the key for that parameter.
...
"""

betatim commented on skopt/searchcv.py:
```diff
@@ -0,0 +1,448 @@
+        key = str(param_space)
+        if key not in self.optimizer_:
+            self.optimizer_[key] = self._make_optimizer(param_space)
+        optimizer = self.optimizer_[key]
```
-> `optimizer = self.optimizer_.get(key, self._make_optimizer(param_space))`

glouppe replied:
Hmm, not sure what is done inside `self._make_optimizer`, but we should ensure there is no border effects... Your suggestion is not equivalent to @iaroslav-ai's code, in case `self._make_optimizer` changes the state of `self` internally.

betatim commented on skopt/searchcv.py:
```diff
@@ -0,0 +1,448 @@
+            n_jobs = max(1, cpu_count() + n_jobs + 1)
+
+        # use the cached optimizer for particular parameter space
+        key = str(param_space)
```
do we need to turn the space into a str before using it as a key? Maybe we should leave a comment where `self.optimizer_` is defined explaining why we can't use `space` directly.

iaroslav-ai replied:
Yeah actually, space is a dict, which is unhashable type, hence I str it. Will add comment.

iaroslav-ai replied:
... and its str seems to be unique for different spaces

betatim replied:
Each optimizer is associated to one of the spaces defined by the user (as a list). So could we make `self.optimizer_` a list as well? Where the order of optimizers corresponds to the order of spaces? Only thought of this yesterday night while pondering how likely it is that we will get subtle bugs because of using `str()` (In my experience constructing a way to hash an unhashable object is the path to pain and subtle bugs)

iaroslav-ai replied:
I would go with your intuition on this, I do not have much experience with hashing in python. Actually some time ago I also had some doubts about this, but forgot about it. I will come up with something that does not require construction of a way to hash a dict with space parameters.

betatim commented:
Having one optimizer for each entry in the list of "spaces" means we can't use the surrogate to choose between the spaces or have some kind of "cross learning" (a word I just made up).

If each subspace had the same dimensions we could simply add an additional categorical dimension to select which sub-space to look at. However we don't have that.

Can't we make one big space out of all the subspaces + one categorical dimension to select which subspace we want and then have the objective select the parameters of interest based on that categorical. Ideally the surrogate would then learn that for certain values of the categorical space the settings of some parameters are irrelevant.

iaroslav-ai commented:
@betatim, I think this is an interesting idea, but I avoided it for the following two reasons:
1. It is a bit easier to implement it as it is right now
2. I am not sure if it will yield better results with converting to one large space for the intended use case of BayesSearchCV.

Explanation for 2:
I assume that different search spaces span parameters for different ML models, and hence are largely independent. Thus concatenation of such spaces has a special structure, which is not assumed by the optimization algorithms in skopt. Thus the algorithm needs to learn such structure accurately, in order to understand that many parameters are independent, and that particular parameter selects different subspaces, and changing parameters in only selected subspace has effect. In the end it might not yield advantage, as subspaces are (assumed to be largely) independent anyway. If they are not, it is unclear how many evaluations it takes for the model to figure this out. 

A different way to see it: let a model parameter space A consists of categorical value a := {1,2,3,4} and b := {4,5,6,7}, and space B of c := {6,7,8,9}. A combination of both has 4 parameters, with total 128 parameter combinations to explore. When explored separately, this results only in 20 combinations to explore. 

Also such search strategy might require large n_calls, in which case fitting surrogate can become rather computationally expensive, eg for GP. 

I propose to consider this later, maybe in context of #108 , where I would actually argue for the approach that you propose. If it is implemented, then we can compare on benchmarks different approaches.

iaroslav-ai commented:
Made a pass over the comments. Please take a look and let me know if you have any other comments.

betatim commented:
Thanks for the long explanation. Sounds sensible. Maybe we can add to the doc string a bit on the separate number of `n_iter` for each subspace and that this gives a "weighting" to each?

Travis is unhappy :(

betatim commented:
Let's see if we can get @glouppe to have a second pass

betatim commented:
There are a few GridSearchCV left over which should be renamed.

betatim commented on skopt/searchcv.py:
```diff
@@ -0,0 +1,425 @@
+import sklearn.model_selection._search as sk_model_sel
+
+import numpy as np
+from collections import defaultdict
```
Can you sort the imports by stdlib, major package, minor package, our package

iaroslav-ai replied:
This is a standard thing to do right? Will do so elsewhere also

betatim commented on skopt/tests/test_searchcv.py:
```diff
@@ -0,0 +1,72 @@
+
+    from sklearn.datasets import load_iris
+    from sklearn.svm import SVC
+    from sklearn.model_selection import train_test_split
```
import these at the top of the file instead of in the function

betatim commented on skopt/utils.py:
```diff
@@ -292,3 +292,75 @@ def cook_estimator(base_estimator, space=None, **kwargs):
     return base_estimator
+
+
+def dims_aslist(params_space):
```
nitpick: spell out params and dims. I'd prefer that despite it being more to type

betatim commented on skopt/utils.py:
```diff
@@ -292,3 +292,75 @@ def cook_estimator(base_estimator, space=None, **kwargs):
+
+def dims_aslist(params_space):
+    """Convert dictionary of parameter names : skopt.space.Dimension
+    a list of skopt.space.Dimension, that can be used in Optimizer.
```
I will think about a different way to formulate this. Right now I had to read it several times despite knowing what the function does :-/

""Convert a dictionary of parameters to a list.

The list of dimensions is created from the values of the
dictionary, sorted by the name of the dimensions used
as keys.

Counterpart to `parameters_asdict`.
"""

Maybe this?

iaroslav-ai replied:
omg lol yeah it was late when I was writing it :D
I will use your version - thanks for coming up with this!

iaroslav-ai commented:
Addressed comments by @betatim . Please let me know if there is something else.

glouppe commented on skopt/__init__.py:
```diff
@@ -61,7 +61,7 @@ def f(x):
 from .utils import dump
 from .utils import expected_minimum
-
+from .searchcv import BayesSearchCV
```
Can you put this in the alphabetical order in the list above?

glouppe commented on skopt/searchcv.py:
```diff
@@ -0,0 +1,425 @@
+import sklearn.model_selection._search as sk_model_sel
+
+from skopt import Optimizer
+from skopt.utils import point_asdict, dimensions_aslist
```
those should be local imports

glouppe commented on skopt/searchcv.py:
```diff
@@ -0,0 +1,425 @@
+    `pre_dispatch` many times. A reasonable value for `pre_dispatch` is `2 *
+    n_jobs`.
+
+    See Also
```
Could we add a short `Example` section in the docstring? (showing that things are in fact simple, despite the huge list of arguments)

glouppe commented on skopt/searchcv.py:
```diff
@@ -0,0 +1,425 @@
+        Either estimator needs to provide a ``score`` function,
+        or ``scoring`` must be passed.
+
+    search_spaces : list of dict or tuple
```
While this is nice, the 90% use case is to search over a single space. I think we should support directly providing a dict. This would also better match with the API of `GridSearchCV`.

glouppe commented on skopt/tests/test_searchcv.py:
```diff
@@ -0,0 +1,67 @@
+
+    # this normally does not hold only if something is wrong
+    # with the optimizaiton procedure as such
+    assert_greater(opt.score(X_test, y_test), 0.9)
```
We should have a test evaluating the multiple search spaces API.

glouppe commented on skopt/searchcv.py:
```diff
@@ -0,0 +1,547 @@
+                            "Parameter of space should be an instance of"
+                            "skopt.space.Dimension (Real, Integer, ...),"
+                            "but in subspace %s the dimension %s is %s" %
+                            (subspace, k, v)
```
Could we instead use `skopt.space.check_dimension`? In this way, we would have a consistent API with `Optimizer/*_minimize` in terms of how dimensions are specified. (Also, this would allow much less boilerplate code for the users.)

iaroslav-ai commented:
This should address all the comments. Will make a second pass over the code later to double check.

glouppe commented:
In the notebook, could you update the minimal example in order to use the simplest API? (directly feeding the dict, with dimensions specified as pairs)

iaroslav-ai commented:
Yup will do so

iaroslav-ai commented:
Updated! :)

glouppe commented:
Thanks! This looks very nice :)

One more thing though... in the last part of the notebook, it seems almost nothing is learned. Within the very first iterations, the optimizer reaches a good value and does not improve from there. Have you tried with more iterations to see if things eventually improve?

iaroslav-ai commented:
I did not try, but my suspicion would be that it will not improve much, as the dataset is not really that complex and is used mainly as a simple example. Maybe I could see what other datasets could be used.

betatim commented on skopt/tests/test_searchcv.py:
```diff
@@ -42,6 +42,15 @@ def test_searchcv_runs(surrogate, n_jobs):
+    def constructor(x): BayesSearchCV(*x)
+    assert_raises(
+        ValueError, constructor, (SVC(), {'C':'1 ... 100.0'})
+    )
```
```python
with pytest.raises(ValueError):
  BayesSearchCV(args_here)
```
https://docs.pytest.org/en/latest/assert.html#assertions-about-expected-exceptions for soem more examples

iaroslav-ai replied:
ahaaa so that is how you do it thx :)

betatim replied:
Trying to spread the pytest way of life :)

betatim commented on skopt/searchcv.py:
```diff
@@ -0,0 +1,547 @@
+                "Search space should be provided as a dict or list of dict,"
+                "got %s" % search_space)
+
+    def add_spaces(self, spaces, names):
```
Not sure it needs to be public, but if it is it needs docs. Which might motivate making it private :)

betatim commented on skopt/utils.py:
```diff
@@ -292,3 +292,99 @@ def cook_estimator(base_estimator, space=None, **kwargs):
+    point_as_list = [
+        point_as_dict[k] for k in sorted(search_space.keys())
+    ]
+    return point_as_list
```
Needs a unit test

betatim replied:
Maybe in `test_utils.py` we can add a test that just tries to round trip a dict to a list and back

MechCoder commented:
Feel free to merge this, I'll try to have a look next week and then make some comments if any. Not sure I will have much to say in any case.

glouppe commented:
@betatim Was this ok with you? I think this PR is good enough to be merged in :) We can polish things further later.

glouppe commented:
:beers: Great work @iaroslav-ai ! Thanks a lot for this, I am pretty sure this will be super useful to many :)

betatim commented:
🍻 and 🍰 !

This is a big new feature! Thanks for the work and patience with the many little comments spread out over days :)

I think we can do a bit more work on improving the doc strings etc but let's do that in a new PR (or several).

iaroslav-ai commented:
A huge thank you to @glouppe  and @betatim for taking your time to review my code! Indeed quite some PR it is :)
I am also testing this internally for some of my things, seems quite useful so far! :)

amueller commented:
Awesome! Did someone do any benchmarks with how this compares to GridSearchCV in particular cases?

betatim commented:
Don't think so. The best we have is https://github.com/scikit-optimize/scikit-optimize/tree/master/benchmarks#ml-classification

Would be great if someone could do perform some benchmarks on "realistic" data sets.

I once saw some benchmarks in https://github.com/rhiever/tpot maybe we can start from there instead of having to do everything ourselves. Not related to tuning ML algos: https://github.com/sigopt/evalset but maybe faster to execute.
